[FLINK-22477] Remove SlotPoolImpl

tillrohrmann opened the pull request:
This PR removes the old `SlotPoolImpl` implementation that was no longer used.

The PR has been split in fine granular commits that first port/remove the old tests and all usages of `SlotPoolImpl`. The second half removes the `SlotPoolImpl` and related methods that were only used by the `SlotPoolImpl`. Please take a look at the individual commits and their commit messages for more context.

zentol reviewed (approved):
+1

zhuzhurk reviewed (commented):
Thanks for opening this PR! @tillrohrmann 
It's really nice cleanup work especially for those legacy tests! I have a few minor comments after reading the first 10+ commits. I will continue with the next commits and will give it a pass through today.

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/benchmark/e2e/SchedulerEndToEndBenchmarkBase.java:
```diff
@@ -74,56 +60,18 @@ public void setup(JobConfiguration jobConfiguration) throws Exception {
-                mainThreadExecutor);
+        final SlotPool slotPool = new DeclarativeSlotPoolBridgeBuilder(mainThreadExecutor).build();
 
         return new PhysicalSlotProviderImpl(
```
Given that declarative slot pool needs to received resource requirements before accepting slot offers, I think we can move the `offerSlots()` into `SchedulingAndDeployingBenchmark#startScheduling()`

tillrohrmann replied:
Ok, will do this change. On a side note, I think the test should guard the assumption that vertices are deployed. With my change I could make it pass even though I broke the underlying assumptions.

zhuzhurk replied:
We will review these benchmarks to see how we can guard them. 
cc @Thesharing

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/PhysicalSlotProviderImplWithSpreadOutStrategyTest.java:
```diff
@@ -109,8 +109,10 @@ public void testSlotAllocationFulfilledWithNewSlots()
+                new DeclarativeSlotPoolBridgeBuilder(
+                                physicalSlotProviderResource.getMainThreadExecutor())
+                        .build();
         assertThat(slotPool.isBatchSlotRequestTimeoutCheckEnabled(), is(true));
```
This is an `org.junit.Assert.assertThat` and can be replaced

tillrohrmann replied:
Will fix it.

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/PhysicalSlotProviderImplWithDefaultSlotSelectionStrategyTest.java:
```diff
@@ -66,8 +66,10 @@ public void testSlotAllocationFulfilledWithNewSlots()
+                new DeclarativeSlotPoolBridgeBuilder(
+                                physicalSlotProviderResource.getMainThreadExecutor())
+                        .build();
         assertThat(slotPool.isBatchSlotRequestTimeoutCheckEnabled(), is(true));
```
This is an `org.junit.Assert.assertThat` and can be replaced

tillrohrmann replied:
Will fix it.

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/SlotPoolUtils.java:
```diff
@@ -55,6 +59,16 @@ private SlotPoolUtils() {
     }
 
+    public static DeclarativeSlotPoolBridge createDeclarativeSlotPoolBridge() {
+        return new DeclarativeSlotPoolBridge(
```
Maybe use `DeclarativeSlotPoolBridgeBuilder` to build it?

tillrohrmann replied:
This won't work because the builder already starts the slot pool bridge which this method does not do. But I think, that I can change the `DeclarativeSlotPoolBridgeBuilder` a bit to make this work.

zhuzhurk reviewed (commented):
I have read through the remaining commits and just three minor comments.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/SlotPoolBatchSlotRequestTest.java:
```diff
@@ -109,30 +106,26 @@ public void testPendingBatchSlotRequestTimeout() throws Exception {
-            final CompletableFuture<PhysicalSlot> firstSlotFuture =
-                    SlotPoolUtils.requestNewAllocatedBatchSlot(
-                            slotPool, directMainThreadExecutor, resourceProfile);
             final CompletableFuture<PhysicalSlot> secondSlotFuture =
```
True, I'll update it.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionGraphRestartTest.java:
```diff
@@ -194,9 +204,27 @@ public void testCancelWhileRestarting() throws Exception {
+        offerSlots(slotPool, new LocalTaskManagerLocation(), numSlots);
+    }
+
+    private void offerSlots(
```
Good point. I'll fix it.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionGraphRestartTest.java:
```diff
@@ -328,18 +361,17 @@ public void testFailingExecutionAfterRestart() throws Exception {
      * Tests that a graph is not restarted after cancellation via a call to {@link
-     * ExecutionGraph#failGlobal(Throwable)}. This can happen when a slot is released concurrently
-     * with cancellation.
+     * ExecutionGraph#failJob(Throwable, long)}. This can happen when a slot is released
```
You are right. I'll fix it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DeclarativeSlotPoolBridge.java:
```diff
@@ -488,6 +488,17 @@ public int getNumPendingRequests() {
     }
 
+    @VisibleForTesting
+    void runBatchSlotTimeoutCheck() {
```
This is a good idea. I'll change it. Moreover, I'll add an `assertRunningInMainThread` to the `checkBatchSlotTimeout` to document that you should not call `checkBatchSlotTimeout` outside of the `mainThreadExectuor`.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/slotpool/SlotPoolBatchSlotRequestTest.java:
```diff
@@ -224,50 +175,16 @@ public void testPendingBatchSlotRequestDoesNotFailIfRMRequestFails() throws Exce
         testingResourceManagerGateway.setRequestSlotFuture(
```
I think this is a good idea. I'll update it.

tillrohrmann commented:
Thanks a lot for the reviews @zentol  and @zhuzhurk. I've addressed all of your comments and pushed an update. I will merge this PR once AZP gives green light.

zhuzhurk reviewed (approved):
+1
